dev-notes/minimal-os-for-fobnail.md

```diff
@@ -0,0 +1,718 @@
+network stack) wouldn't compromise the entire OS (contrary to monolithic
+kernels).
+
+We have opened an [issue](https://github.com/seL4/seL4/issues/832) about seL4
+usage as secure bootloader.
+
+#### Running in DLME
+
+It should be runnable, however due to microkernel nature all programs run in
+unprivileged mode which complicates booting of the target OS. seL4 would have to
+gain kexec-like abilities. seL4 developers are
+[willing](https://github.com/seL4/seL4/issues/832#issuecomment-1114385689) to
+add kexec support.
+
+#### Fobnail integration
+
+seL4 provides virtually no drivers, except a few drivers listed
+[here](https://docs.sel4.systems/projects/available-user-components.html#device-drivers).
+According to the page linked above, `libusbdrivers` is inactive and lacks XHCI
+support. seL4 has basic support for an old version of musl C library (v1.1.16).
+
+Using seL4 would require a significant amount of work
+- Extending USB drivers
+- Implementing USB EEM driver
+- Implementing TPM driver
+- Porting [Fobnail Attester](https://github.com/fobnail/fobnail-attester)
+- seL4 runs all processes in unprivileged mode, so the kernel itself would have
+  to be modified to allow booting of another kernel
+
+[Genode](https://github.com/genodelabs/genode) can provide many of the required
+drivers. See the section below for details.
+
+CAmkES provides tools for writing seL4-native components, it also has VMM
+support for running virtual machines. Contrary to Genode CAmkES does not provide
+drivers, but VMs may provide them. However this requires virtualization and seL4
+currently supports it only on Intel CPUs. See the section below for more details.
+
+### Genode
+
+[Genode](https://github.com/genodelabs/genode) is a framework for creating
+custom, specialized OSes. It can work on various kernels/hypervisors, including
+seL4 and Linux. It provides many drivers, including USB drivers (including
+`usb-net` driver which should handle USB EEM) and a network stack. When running
+on Linux all components are sandboxed with seccomp, however it should be noted
+that Linux drivers are used. For other kernels Genode provides sandboxed
+drivers.
+
+#### Running in DLME
+
+Whether Genode can run in DLME depends on the choosen kernel, however we are
+interested mainly in seL4. Please see seL4 section above for details.
+
+#### Fobnail integration
+
+TPM driver would have to be ported to Genode. Underlying kernel must have kexec
+abilities and Genode must have component capable of invoking kexec. USB and
+EEM drivers should work out-of-the-box.
+
+Genode's seL4 support is considered to be in healthly state (see
+[this](https://github.com/genodelabs/genode/issues/4480#issuecomment-1108525737)
+comment). But it should be noted Genode with seL4 has not been used outside the
+lab so far.
+
+Also, it should be noted that Genode license (AGPLv3) is very restrictive and it
+could be a problem for the Fobnail Project.
```

Comment:
Why in other cases we don't talk about licenses?

Reply:
In this case we added it from the review section (https://github.com/fobnail/docs/pull/26#discussion_r857764027). Should we add license as hard or soft requirements and think about them in other OSes?

Reply:
@TomaszAIR always, every software analysis should consider or at least mention license. This is fundamental part of software development craftsmanship.

Reply:
Licenses added in [0157aad](https://github.com/fobnail/docs/pull/26/commits/0157aad6fef7b0974587b919f0748211b5e7dc91) for every system and to the table as a requirement, score added in [3f13874](https://github.com/fobnail/docs/pull/26/commits/3f1387424ee2a95ac3d6c909a585cb146dc3349e)